### PR TITLE
Remove safe_yaml.

### PR DESCRIPTION
**Pull request by bf4**
I don't think it's really this gem's job, but besides that, I was getting a failure
when starting zeus, so I just made it optional.

The below stacktrace is from a fork of v0.6.7 with a commit on top of it:

https://github.com/swipesense/rails_admin/commits/0.6.7-eager_loading

Here's the stacktrace:

```plain
$ zeus rake

     ~/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/psych/parser.rb:33:in `<class:Parser>': superclass mismatch for class Mark (TypeError)
from ~/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/psych/parser.rb:32:in `<module:Psych>'
from ~/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/psych/parser.rb:1:in `<top (required)>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `block in require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:240:in `load_dependency'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/psych.rb:7:in `<top (required)>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `block in require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:240:in `load_dependency'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/safe_yaml-1.0.4/lib/safe_yaml/psych_handler.rb:1:in `<top (required)>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `block in require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:240:in `load_dependency'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/safe_yaml-1.0.4/lib/safe_yaml/load.rb:130:in `<module:SafeYAML>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/safe_yaml-1.0.4/lib/safe_yaml/load.rb:26:in `<top (required)>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `block in require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:240:in `load_dependency'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/safe_yaml-1.0.4/lib/safe_yaml.rb:1:in `<top (required)>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `block in require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:240:in `load_dependency'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/bundler/gems/rails_admin-975bb5075824/lib/rails_admin/engine.rb:10:in `<top (required)>'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `block in require'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:240:in `load_dependency'
from ~/projects/rails_server/bundle/ruby/2.2.0/gems/activesupport-4.2.3/lib/active_support/dependencies.rb:274:in `require'
from ~/projects/rails_server/bundle/ruby/2.2.0/bundler/gems/rails_admin-975bb5075824/lib/rails_admin.rb:1:in `<top (required)>'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:76:in `require'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:76:in `block (2 levels) in require'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:72:in `each'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:72:in `block in require'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:61:in `each'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:61:in `require'
from ~/.rvm/gems/ruby-2.2.2/gems/bundler-1.10.6/lib/bundler.rb:134:in `require'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus/rails.rb:97:in `default_bundle'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus.rb:200:in `run_action'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus.rb:74:in `block (2 levels) in boot_steps'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus/load_tracking.rb:7:in `features_loaded_by'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus.rb:73:in `block in boot_steps'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus.rb:56:in `catch'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus.rb:56:in `boot_steps'
from ~/.rvm/gems/ruby-2.2.2/gems/zeus-0.15.4/lib/zeus.rb:48:in `go'
from -e:1:in `<main>'
```

### CODE REVIEW

**Comment by bf4** [comment]
Rebased against current master

File: lib/rails_admin/engine.rb [hunk 1]
```diff
@@ -7,9 +7,28 @@
...
+    require 'safe_yaml/load'
+  rescue LoadError
+    if YAML.respond_to?(:safe_load)
+      SafeYAML = Module.new do
+        def self.load(yaml)
+          YAML.safe_load(yaml)
+        end
+      end
+    else
+      warn "Safe-loading of YAML is not available. Please install 'safe_yaml' or install Psych 2.0+"
```
Comment (mshibuya) [inline lib/rails_admin/engine.rb]: An error should be raised here, not warning. We shouldn't let users choose unsafe state.

Comment (bf4) [inline lib/rails_admin/engine.rb]: Will change

B mobile phone

File: lib/rails_admin/engine.rb [hunk 2]
```diff
@@ -7,9 +7,22 @@
...
+require 'yaml'
 
 module RailsAdmin
+  # Backwards-compatible RailsAdmin::SafeYAML when SafeYAML isn't available
+  begin
+    require 'safe_yaml/load'
+  rescue LoadError
+    unless YAML.respond_to?(:safe_yaml)
+      raise LoadError, "Safe-loading of YAML is not available. Please install 'safe_yaml' or install Psych 2.0+"
+    end
```
Comment (bf4) [inline lib/rails_admin/engine.rb]: Guard clause raises same exception but with better message when there's no fallback

File: lib/rails_admin/engine.rb [hunk 3]
```diff
@@ -7,9 +7,22 @@
...
 require 'remotipart'
-require 'safe_yaml/load'
+require 'yaml'
 
 module RailsAdmin
+  # Backwards-compatible RailsAdmin::SafeYAML when SafeYAML isn't available
+  begin
+    require 'safe_yaml/load'
+  rescue LoadError
+    unless YAML.respond_to?(:safe_yaml)
```
Comment (bf4) [inline lib/rails_admin/engine.rb]: typo: should be `respond_to?(:safe_load)` will fix

File: lib/rails_admin/engine.rb [hunk 4]
```diff
@@ -7,9 +7,22 @@
 require 'rails'
 require 'rails_admin'
 require 'remotipart'
-require 'safe_yaml/load'
+require 'yaml'
 
 module RailsAdmin
+  # Backwards-compatible RailsAdmin::SafeYAML when SafeYAML isn't available
+  begin
+    require 'safe_yaml/load'
```
Comment (bf4) [inline lib/rails_admin/engine.rb]: Missing: `SafeYAML = ::SafeYAML`

**Comment by bbenezech** [comment]
@bf4 can we merge this?

**Comment by bf4** [comment]
@bbenezech I think it can be merged, and is actually better since you last asked the question.

File: lib/rails_admin.rb [hunk 5]
```diff
@@ -33,6 +34,28 @@ def self.config(entity = nil, &block)
...
   end
+
+  # Backwards-compatible with safe_yaml/load when SafeYAML isn't available.
+  # Evaluates available YAML loaders at boot and creates appropriate method,
+  # so no conditionals are required at runtime.
+  begin
+    require 'safe_yaml/load'
+    def self.yaml_load(yaml)
+      SafeYAML.load(yaml)
+    end
```
Comment (bf4) [inline lib/rails_admin.rb]: define `self.yaml_load` to use SafeYAML in non-monkey-patching mode, when available

File: lib/rails_admin.rb [hunk 6]
```diff
@@ -33,6 +34,28 @@ def self.config(entity = nil, &block)
...
+  rescue LoadError
+    if YAML.respond_to?(:safe_load)
```
Comment (bf4) [inline lib/rails_admin.rb]: define `self.yaml_load` to use Psych.safe_load when available

File: lib/rails_admin.rb [hunk 7]
```diff
@@ -33,6 +34,28 @@ def self.config(entity = nil, &block)
...
+      def self.yaml_load(yaml)
+        YAML.safe_load(yaml)
+      end
+    else
+      raise LoadError.new "Safe-loading of YAML is not available. Please install 'safe_yaml' or install Psych 2.0+"
+    end
```
Comment (bf4) [inline lib/rails_admin.rb]: Fail with useful warning if user does not have safe_yaml or Psych 2.0+

File: lib/rails_admin.rb [hunk 8]
```diff
@@ -33,6 +34,28 @@ def self.config(entity = nil, &block)
       RailsAdmin::Config
     end
   end
+
+  # Backwards-compatible with safe_yaml/load when SafeYAML isn't available.
+  # Evaluates available YAML loaders at boot and creates appropriate method,
+  # so no conditionals are required at runtime.
```
Comment (bf4) [inline lib/rails_admin.rb]: two conditions for defining `RailsAdmin.yaml_load` depending on available gems. But the nicest part is, that this method can be customized by any end user :)

**Comment by bbenezech** [comment]
Thanks a lot. It's good to see safe_yaml go. (with psych 2 at least)

**Comment by bf4** [comment]
Awesome!

**Pull request merged**